plugins/Login/Login.php

```diff
@@ -33,13 +38,76 @@ public function registerEvents()
+            'Login.authenticate.successful'    => 'beforeLoginCheckBruteForce',
+            'Login.authenticate.failed'        => 'onFailedLoginRecordAttempt', // record any failed attempt in UI
+            'API.Request.authenticate.failed' => 'onFailedLoginRecordAttempt', // record any failed attempt in Reporting API
+            'Tracker.Request.authenticate.failed' => 'onFailedLoginRecordAttempt', // record any failed attempt in Tracker API
         );
         return $hooks;
     }
 
+    public function isTrackerPlugin()
+    {
+        return true;
+    }
+
+    public function onInitAuthenticationObject()
+    {
+        if (SettingsServer::isTrackerApiRequest() || Request::isRootRequestApiRequest()) {
+            // we check it for all API requests...
+            // we do not check it for other UI requests as otherwise we would be logging out someone possibly already
+            // logged in with a valid session which we don't want currently... regular UI requests are checked through
+            // 1) any successful or failed login attempt, plus through specific controller action that a user can use
+            // to log in
+            $this->beforeLoginCheckBruteForce();
+        }
+    }
+
+    public function onFailedLoginRecordAttempt()
+    {
+        // we're always making sure on any success or failed login to check if user is actually allowed to log in
+        // in case for some reason it forgot to run the check
+        $this->beforeLoginCheckBruteForce();
+
+        // we are recording new failed attempts only when user can currently log in and is not blocked...
+        // this is to kind of block eg a certain IP continuously. could alternatively also still keep writing those failed
+        // attempts into the log and only allow login attempts again after the user had no login attempts for the configured
+        // time frame
+        $bruteForce = StaticContainer::get('Piwik\Plugins\Login\Security\BruteForceDetection');
+        if ($bruteForce->isEnabled()) {
+            if (!$this->hasAddedFailedAttempt) {
+                $bruteForce->addFailedLoginAttempt(IP::getIpFromHeader());
+                // we make sure to log max one failed login attempt per request... otherwise we might log 3 or many more
+                // if eg API is called etc.
+                $this->hasAddedFailedAttempt = true;
+            }
+        }
+    }
+
+    public function beforeLoginCheckBruteForce()
+    {
+        $bruteForce = StaticContainer::get('Piwik\Plugins\Login\Security\BruteForceDetection');
+        if (!$this->hasPerformedBruteForceCheck && $bruteForce->isEnabled() && !$bruteForce->canLogin(IP::getIpFromHeader())) {
+            throw new Exception(Piwik::translate('Login_LoginNotAllowedBecauseBlocked'));
+        }
+        // for performance reasons we make sure to execute it only once per request
+        $this->hasPerformedBruteForceCheck = true;
```

Comment:
If `onFailedLoginRecordAttempt()` is called in a request, then this method can get called more than once, correct? Perhaps we should check if `$this->hasPerformedBruteForceCheck` is true at the top of this method and return if so?

Reply:
not sure I understand what you mean... it already checks this flag in the `if` in `beforeLoginCheckBruteForce()`.  and we have a different flag in`onFailedLoginRecordAttempt`

Reply:
Nvm, missed the `if`.